docs/reference/migration/apis/data-stream-reindex.asciidoc

```diff
@@ -0,0 +1,358 @@
+--
+
+include::{es-ref-dir}/migration/apis/shared-migration-apis-tip.asciidoc[]
+
+The reindex data stream API is used to upgrade the backing indices of a data stream to the most
+recent major version. It works by reindexing each backing index then replacing the original
+backing index with its replacement and deleting the original backing index. The settings and mappings
+from the original backing indices are copied to the resulting backing indices.
+
+This api runs asynchronously because reindexing all indices in a large data stream
+is expected to take a large amount of time and resources. The endpoint will return immediately and a persistent
+task will be created to run in the background. The current status of the task can be checked with
+the <<data-stream-reindex-status-api,reindex status API>>. This status will be available for 24 hours after the task completes, whether
+it finished successfully or failed. A running data stream reindex task can be cancelled using the
+<<data-stream-reindex-cancel-api,reindex cancel API>>.
+
+[source,console]
+----
+POST _migration/reindex
+{
+    "source": {
+        "index": "my-data-stream"
+    },
+    "mode": "upgrade"
+}
+----
+// TEST[setup:my_data_stream]
+// TEST[teardown:data_stream_cleanup]
+
+
+
+[[data-stream-reindex-api-request]]
+==== {api-request-title}
+
+`POST /_migration/reindex`
+
+
+[[data-stream-reindex-api-prereqs]]
+==== {api-prereq-title}
+
+* If the {es} {security-features} are enabled, you must have the `manage`
+<<privileges-list-indices,index privilege>> for the data stream.
+
+[[data-stream-reindex-body]]
+==== {api-request-body-title}
+
+`source`::
+`index`:::
+(Required, string) The name of the data stream to upgrade.
+
+`mode`::
+(Required, enum) Set to `upgrade` to upgrade the data stream in-place, using the same source and destination
+data stream. Each backing index will be reindexed, then swapped into the data stream, then the old index will be deleted.
+Currently, the only allowed value for this parameter is `upgrade`.
+
+[[reindex-data-stream-api-settings]]
+==== Settings
+
+You can use the following settings to control the behavior of the reindex data stream API:
+
+[[migrate_max_concurrent_indices_reindexed_per_data_stream-setting]]
+// tag::migrate_max_concurrent_indices_reindexed_per_data_stream-setting-tag[]
+`migrate.max_concurrent_indices_reindexed_per_data_stream`
+(<<dynamic-cluster-setting,Dynamic>>)
+The number of backing indices within a given data stream which will be reindexed concurrently. Defaults to `1`.
+// end::migrate_max_concurrent_indices_reindexed_per_data_stream-tag[]
+
+[[migrate_data_stream_reindex_max_request_per_second-setting]]
+// tag::migrate_data_stream_reindex_max_request_per_second-setting-tag[]
+`migrate.data_stream_reindex_max_request_per_second`
+(<<dynamic-cluster-setting,Dynamic>>)
+The average number of documents within a given backing index to reindex per second.
```

Comment:
maybe `The average **maximum** number of documents within a given backing index to reindex per second.`? Otherwise, I'd just set it to infinity to get infinite throughput! :)